### PR TITLE
Move cuda kernels to their own file and namespace

### PR DESCRIPTION
**Pull request by Rombur**
Move the cuda kernel out of cuda_vector.

### CODE REVIEW

File: include/deal.II/lac/cuda_kernel.h [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+#ifndef dealii_cuda_kernel_h
+#define dealii_cuda_kernel_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/cuda_size.h>
+
+#include <deal.II/lac/cuda_atomic.h>
+#include <deal.II/lac/cuda_vector.h>
+
```
Comment (jppelteret) [inline include/deal.II/lac/cuda_kernel.h]: Does this not need an `#ifdef DEAL_II_WITH_CUDA` here, as is done in `lac/cuda_vector.cu`?

Comment (Rombur) [reply include/deal.II/lac/cuda_kernel.h]: Yes, I actually cherry-pick this from another branch I am working on and I have added the preprocessor guard in another commit. Thanks.

**Comment by Rombur** [comment]
@masterleinad This is a good remark. I will reuse these kernels several times that's why I moved them. However, we have a bunch of similar functions in `vector_operations_internal` that don't have doxygen documentation and are all in one file. Because the cuda kernels are similar to the cpu functions, I didn't bother writing doxygen documentation. If you want, I can write a proper documentation but I am not sure if there is an interest for it.

**Comment by jppelteret** [comment]
> we should have doxygen documentation and maybe separate the declarations from the definitions as we usually do

I mirror @masterleinad's opinion (I thought about mentioning similar in my initial comment)...

>  If you want, I can write a proper documentation but I am not sure if there is an interest for it.

... but I think that maybe we can open an issue for this and we can add the documentation in another PR. Would that be a reasonable compromise?

**Comment by masterleinad** [comment]
> ... but I think that maybe we can open an issue for this and we can add the documentation in another PR. Would that be a reasonable compromise?

Yes, I thought about suggesting that as well.

**Comment by Rombur** [comment]
> ... but I think that maybe we can open an issue for this and we can add the documentation in another PR. Would that be a reasonable compromise?

No, that's fine there are others changes to this file that I was planning to push later. I will just consolidate everything in this PR.

**Comment by Rombur** [comment]
Ready for re-review.

**Comment by drwells** [comment]
@Rombur Congratulations on making the 42,000th commit to deal.II!

**Pull request merged**